build-tools/cmake/xa_macros.cmake

```diff
@@ -77,3 +77,43 @@ macro(xa_common_prepare)
       )
   endif()
 endmacro()
+
+macro(xa_macos_prepare_arm64)
+   if(APPLE)
+     if(CMAKE_CXX_COMPILER_VERSION)
+       set(COMPILER_VERSION ${CMAKE_CXX_COMPILER_VERSION})
+     endif()
+
+     if(CMAKE_C_COMPILER_VERSION AND NOT COMPILER_VERSION)
+       set(COMPILER_VERSION ${CMAKE_C_COMPILER_VERSION})
+     endif()
+
+     unset(COMPILER_VERSION_MAJOR)
+     if(COMPILER_VERSION)
+        message(STATUS "macOS C/C++ compiler version == ${COMPILER_VERSION}")
+        string(REPLACE "." ";" COMPILER_VERSION_LIST ${COMPILER_VERSION})
+        list(GET COMPILER_VERSION_LIST 0 COMPILER_VERSION_MAJOR)
+        message(STATUS "major compiler version: ${COMPILER_VERSION_MAJOR}")
+     else()
+        message(WARNING "Unable to detect macOS C/C++ compiler version")
+     endif()
+
+     set(COMPILER_SUPPORTS_ARM64 False)
+     if(COMPILER_VERSION_MAJOR)
+        if(${COMPILER_VERSION_MAJOR} GREATER_EQUAL 12)
```

Comment:
A possible problem here is that we're assuming that if the compiler version is >= 12, then `-arch arm64` is supported.

The problem is that there are *two* Xcode 12 downloads:

  * **Xcode 12 beta 2**: 
    > This distribution of Xcode 12 beta does not include support for building macOS Universal apps. 
  * **Xcode 12 for macOS Universal Apps beta 2**:
    > This version of Xcode 12 beta is intended exclusively for building macOS Universal apps that run on Apple Silicon Macs

I haven't downloaded the former to verify that it reports a compiler version of 12, but I imagine it *would* report a compiler version of 12 while *not* supporting `clang -arch arm64`.

Is there a better/"more robust" way of determining that `clang -arch arm64` is supported?

Reply:
Then again…?

My `/Applications/Xcode.app` is Xcode 11 (stable), with `clang --version` reporting:

```
Apple clang version 11.0.3 (clang-1103.0.32.62)
```

Does Xcode 11 support `-arch arm64`?  Kinda!

```bash
$ /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang -o h hw.c -arch arm64 \
-isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/usr/include/sys/cdefs.h:807:2: error: Unsupported architecture
#error Unsupported architecture
 ^
```

Note that this isn't from the compiler; it's from a header file.  If we use `clang -isysroot` using an Xcode 12 path, it "works":

```bash
$ /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang -o h hw.c -arch arm64 \
    -isysroot /Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.0.sdk
ld: warning: ignoring file /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/11.0.3/lib/darwin/libclang_rt.osx.a, missing required architecture arm64 in file /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/11.0.3/lib/darwin/libclang_rt.osx.a (3 slices)
$ echo $?
0
```

It works in that the exit code is 0.  There is a warning about `libclang_rt.osx.a`, but the file *is* created, and contains arm64 output:

```bash
$ file h
h: Mach-O 64-bit executable arm64
```

Which is somewhat bizarre.

Reply:
Which is all to suggest that maybe a "`clang` version check" is the wrong course of action here, and perhaps instead check for a *file* within the SDK which suggests that arm64 is supported?

Swift documentation appears to be per-ABI, resulting in files such as:

```
/Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.0.sdk/System/iOSSupport/usr/lib/swift/CoreNFC.swiftmodule/arm64.swiftdoc
```

Xcode 11 has a `CoreNFC.swiftmodule/x86_64.swiftdoc`, but no `arm64.swiftdoc` file.

Alternatively, the `SDKSettings.json` file mentions it:

```
/* /Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.0.sdk//SDKSettings.json */
…
"DeploymentTargetSettingName":"MACOSX_DEPLOYMENT_TARGET","Archs":["x86_64","x86_64h","arm64"]
…
```

It's all one line, but the Xcode 11 `SDKSettings.json` file doesn't contain *any* mention of arm64`, meaning `grep` may be sufficient:

```
$ grep arm64 /Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.0.sdk//SDKSettings.json 2>&1 > /dev/null && echo yes
yes
```